regression 1.49: problem with autoderef and trait method

A code attached below used to compile with rustc 1.48 (stable) and stopped to work with 1.49 (stable). Two thing help to make the code compile:
1) comment out `use arc_swap::access::Access;` 
2) explicitly deref `(*s).load()`. 

To compile code below one need to add `arc-swap = "1.2"` to Cargo.toml dependencies. Maybe this is due to some conflict between ArcSwap instance method and Access load method?

### Code

I tried this code:

```rust
use arc_swap;
use std::sync::Arc;
use arc_swap::access::Access;

pub fn foo() {
    let s: Arc<arc_swap::ArcSwap<usize>> = Arc::new(arc_swap::ArcSwap::from_pointee(32))
    let guard: arc_swap::Guard<Arc<usize>> = s.load();
}
```

I expected to see this happen: It should compile properly (it worked with rustc version 1.48 and earlier versions as well).

Instead, this happened: I got compilation error
```
error[E0283]: type annotations needed
 --> src/lib.rs:9:50
  |
9 |     let guard: arc_swap::Guard<Arc<usize>> =   s.load();
  |                                                  ^^^^ cannot infer type for type parameter `T`
  |
  = note: cannot satisfy `ArcSwapAny<Arc<usize>>: arc_swap::access::Access<_>`
  = note: required because of the requirements on the impl of `arc_swap::access::Access<_>` for `Arc<ArcSwapAny<Arc<usize>>>`
```

### Version it worked on

It most recently worked on: Rust 1.48

### Version with regression

`rustc --version --verbose`:
```
rustc 1.49.0 (e1884a8e3 2020-12-29)
binary: rustc
commit-hash: e1884a8e3c3e813aada8254edfa120e85bf5ffca
commit-date: 2020-12-29
host: x86_64-unknown-linux-gnu
release: 1.49.0
```

--- a/src/test/ui/inference/issue-80816.rs
+++ b/src/test/ui/inference/issue-80816.rs
@@ -1,0 +1,54 @@
+#![allow(unreachable_code)]
+
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::sync::Arc;
+
+pub struct Guard<T> {
+    _phantom: PhantomData<T>,
+}
+impl<T> Deref for Guard<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        unimplemented!()
+    }
+}
+
+pub struct DirectDeref<T>(T);
+impl<T> Deref for DirectDeref<Arc<T>> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        unimplemented!()
+    }
+}
+
+pub trait Access<T> {
+    type Guard: Deref<Target = T>;
+    fn load(&self) -> Self::Guard {
+        unimplemented!()
+    }
+}
+impl<T, A: Access<T>, P: Deref<Target = A>> Access<T> for P {
+    //~^ NOTE: required for `Arc<ArcSwapAny<Arc<usize>>>` to implement `Access<_>`
+    type Guard = A::Guard;
+}
+impl<T> Access<T> for ArcSwapAny<T> {
+    //~^ NOTE: multiple `impl`s satisfying `ArcSwapAny<Arc<usize>>: Access<_>` found
+    type Guard = Guard<T>;
+}
+impl<T> Access<T> for ArcSwapAny<Arc<T>> {
+    type Guard = DirectDeref<Arc<T>>;
+}
+
+pub struct ArcSwapAny<T> {
+    _phantom_arc: PhantomData<T>,
+}
+
+pub fn foo() {
+    let s: Arc<ArcSwapAny<Arc<usize>>> = unimplemented!();
+    let guard: Guard<Arc<usize>> = s.load();
+    //~^ ERROR: type annotations needed
+    //~| HELP: try using a fully qualified path to specify the expected types
+}
+
+fn main() {}

--- a/src/test/ui/inference/issue-80816.stderr
+++ b/src/test/ui/inference/issue-80816.stderr
@@ -1,0 +1,27 @@
+error[E0283]: type annotations needed
+  --> $DIR/issue-80816.rs:49:38
+   |
+LL |     let guard: Guard<Arc<usize>> = s.load();
+   |                                      ^^^^
+   |
+note: multiple `impl`s satisfying `ArcSwapAny<Arc<usize>>: Access<_>` found
+  --> $DIR/issue-80816.rs:35:1
+   |
+LL | impl<T> Access<T> for ArcSwapAny<T> {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+...
+LL | impl<T> Access<T> for ArcSwapAny<Arc<T>> {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+note: required for `Arc<ArcSwapAny<Arc<usize>>>` to implement `Access<_>`
+  --> $DIR/issue-80816.rs:31:45
+   |
+LL | impl<T, A: Access<T>, P: Deref<Target = A>> Access<T> for P {
+   |                                             ^^^^^^^^^     ^
+help: try using a fully qualified path to specify the expected types
+   |
+LL |     let guard: Guard<Arc<usize>> = <Arc<ArcSwapAny<Arc<usize>>> as Access<T>>::load(&s);
+   |                                    ++++++++++++++++++++++++++++++++++++++++++++++++++ ~
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0283`.

